File: script/robot_control.py
import rospy
from sensor_msgs.msg import JointState
from std_msgs.msg import Float64MultiArray
import numpy as np
from robot_control_solution.robot_kinematics import UR5Kinematics
from ruckig import InputParameter, OutputParameter, Result, Ruckig
from dataclasses import dataclass

@dataclass
class MotionProfileSettings:
    max_velocity: float  # Maximum velocity
    max_acceleration: float  # Maximum acceleration
    jerk: float  # Maximum jerk

@dataclass
class CartMotionProfileSettings:
    angular: MotionProfileSettings  # Angular motion profile settings
    linear: MotionProfileSettings  # Linear motion profile settings


class UR5Control(UR5Kinematics): 
    def __init__(self, freq: int = 100, joints_profile: MotionProfileSettings = None, 
                 cart_profile: CartMotionProfileSettings = None):
        """
        Initialize the UR5Control class.

        :param freq: Frequency of control loop in Hz.
        :param joints_profile: Motion profile settings for joint movement.
        :param cart_profile: Motion profile settings for Cartesian movement.
        """
        super().__init__()

        # Set default motion profile settings if not provided
        if joints_profile is None:
            joints_profile = MotionProfileSettings(0.3, 1.0, 1.0)  # max_velocity, max_acceleration, jerk
        
        if cart_profile is None:
            cart_profile = CartMotionProfileSettings(
                MotionProfileSettings(0.1, 0.2, 0.3),  # Angular motion settings
                MotionProfileSettings(0.1, 0.2, 0.3)   # Linear motion settings
            )
        
        # Class attributes initialization
        self.__freq: int = freq  # Control loop frequency
        self.__current_joints_vel: np.ndarray = np.zeros(6)  # Current joint velocities
        self.__current_joints_pos: np.ndarray = np.zeros(6)  # Current joint positions
        self.__joints_last_pos: np.ndarray = np.zeros(6)
        self.__joints_profile: MotionProfileSettings = joints_profile  # Joint motion profile settings
        self.__cart_profile: CartMotionProfileSettings = cart_profile  # Cartesian motion profile settings

        # Subscribe to joint state updates and set up publishers
        self.joint_state_subscriber = rospy.Subscriber("/joint_states", JointState, self.jointStateCallback)
        self.joint_position_pub = rospy.Publisher("/joint_group_position_controller/command", Float64MultiArray, queue_size=10)

    def jointStateCallback(self, msg: JointState) -> None:
        """
        Callback for updating joint states from incoming messages.

        :param msg: The joint state message received.
        """
        if len(msg.position) == 6:  
            # Update the current joint velocities and positions
            self.__current_joints_vel = np.array(msg.velocity)
            self.__current_joints_pos = np.array(msg.position)  
        else:
            rospy.logwarn("Received joint state with insufficient position data.")

    def readJointsStates(self):
        """
        Read the current joint positions and velocities.

        :return: A tuple containing current joint positions and velocities.
        """
        return self.__current_joints_pos, self.__current_joints_vel
    
    def setJointPos(self, joints_pos: np.ndarray) -> None:
        """
        Publish the desired joint positions to the joint position topic.

        :param joints_pos: Array of target joint positions.
        """
        joint_positions = Float64MultiArray()
        joint_positions.data = list(joints_pos)
        self.__joints_last_pos = joints_pos
        self.joint_position_pub.publish(joint_positions)

    def movePtP(self, target_transform: np.ndarray, profile_settings: MotionProfileSettings = None) -> bool:
        """
        Move the robot in a point-to-point motion to the specified target transformation.

        :param target_transform: The target transformation matrix.
        :param profile_settings: Optional motion profile settings for the movement.
        :return: True if the movement was successful, False otherwise.
        """
        try:
            if profile_settings is None:
                profile_settings = self.__joints_profile

            # Obtain possible joint solutions for the target transformation
            solutions = np.empty((8, 6), dtype=float)
        
            num_solutions = self.inverseKinematics(target_transform, solutions)
            if num_solutions == 0:
                rospy.logerr("Target position is not reachable")
                return False

            # Find the nearest joint solution for the desired target
            target_joints_angles = self.calculateNearestIKSolution(target_transform, self.__joints_last_pos)
            
            # Initialize Ruckig for 6 joints
            otg = Ruckig(6, 1.0/self.__freq)
            input = InputParameter(6)  # Input parameter for 6 joints
            output = OutputParameter(6)  # Output parameter for 6 joints

            # Set initial conditions
            input.current_velocity = np.zeros(6)
            input.current_acceleration = np.zeros(6)
            input.target_velocity = np.zeros(6)
            input.target_acceleration = np.zeros(6)
            input.current_position = self.__joints_last_pos  # Current joint angles
            input.target_position = target_joints_angles  # Target joint angles

            # Set maximum motion parameters
            input.max_velocity = np.full(6, self.__joints_profile.max_velocity)
            input.max_acceleration = np.full(6, self.__joints_profile.max_acceleration)
            input.max_jerk = np.full(6, self.__joints_profile.jerk)

            # Update position loop
            rate = rospy.Rate(self.__freq)
            res = Result.Working
            while (res == Result.Working) and (not rospy.is_shutdown()):
                res = otg.update(input, output)
                # Set new joint angles
                self.setJointPos(np.array(output.new_position))
                output.pass_to_input(input)  # Update input parameters
                rate.sleep()  # Sleep to maintain the control frequency
            return True
        except Exception as e:
            rospy.logerr(e)
            return False
    
    def moveLine(self, target_transform: np.ndarray, profile_settings: CartMotionProfileSettings = None) -> bool:
        """
        Move the robot in a linear trajectory to the specified target transformation.

        :param target_transform: The target transformation matrix.
        :param profile_settings: Optional motion profile settings for the movement.
        :return: True if the movement was successful, False otherwise.
        """
        try:
            if profile_settings is None:
                profile_settings = self.__cart_profile
                
            # Get the current joint angles and convert them to a transformation matrix
            current_transform = self.forwardKinematics(self.__joints_last_pos)

            # Extract quaternions and positions using methods from UR5Kinematics
            current_quat, current_pos = self.extractRotationAndTranslation(current_transform)
            target_quat, target_pos = self.extractRotationAndTranslation(target_transform)

            # Obtain solutions for the target pose
            target_solutions = np.empty((8, 6), dtype=float)
            num_solutions = self.inverseKinematics(target_transform, target_solutions)
            if num_solutions == 0:
                rospy.logerr("Target position is not reachable")
                return False

            # Initialize Ruckig for movement
            input = InputParameter(7)  # Input parameter for 7 elements (4 for quaternion, 3 for position)
            output = OutputParameter(7)  # Output parameter for 7 elements
            otg = Ruckig(7, 1.0/self.__freq)  # Create Ruckig instance with the specified frequency

            # Set initial conditions
            input.current_velocity = np.zeros(7)
            input.current_acceleration = np.zeros(7)
            input.target_velocity = np.zeros(7)
            input.target_acceleration = np.zeros(7)

            # Set current and target positions
            input.current_position = np.concatenate((current_quat, current_pos))
            input.target_position = np.concatenate((target_quat, target_pos))

            input.max_velocity = np.empty(7, dtype=float)
            input.max_acceleration = np.empty(7, dtype=float)
            input.max_jerk = np.empty(7, dtype=float)

            input.max_velocity = np.concatenate((np.full(4,self.__cart_profile.angular.max_velocity), np.full(3,self.__cart_profile.linear.max_velocity)))
            input.max_acceleration = np.concatenate((np.full(4,self.__cart_profile.angular.max_acceleration), np.full(3,self.__cart_profile.linear.max_acceleration)))
            input.max_jerk = np.concatenate((np.full(4,self.__cart_profile.angular.jerk), np.full(3,self.__cart_profile.linear.jerk)))

            rate = rospy.Rate(self.__freq)
            res = Result.Working
            while (res == Result.Working) and (not rospy.is_shutdown()):
                res = otg.update(input, output)
                # Extract new positions and convert them for inverse kinematics
                new_quat = output.new_position[:4]
                new_pos = output.new_position[4:]
                
                new_quat /= np.linalg.norm(new_quat)
                new_transform = self.createTransformationMatrix(new_quat, new_pos)
                target_solutions = np.empty((8, 6), dtype=float)
                num_solution = self.inverseKinematics(new_transform, target_solutions)
                if num_solution == 0:
                    rospy.logerr("Target position is not reachable")
                    return False
                
                # Find the nearest joint solution
                target_joints_angles = self.calculateNearestIKSolution(new_transform, self.__joints_last_pos)
                self.setJointPos(target_joints_angles)
                
                output.pass_to_input(input)  # Update input parameters
                rate.sleep()
            return True
        except Exception as e:
            rospy.logerr(e)
            return False
    
    def extractRotationAndTranslation(self, transformation: np.ndarray) -> tuple:
        """
        Extracts the rotation (as a quaternion) and translation from a transformation matrix.

        :param transformation: A 4x4 transformation matrix.
        :return: A tuple containing the quaternion and translation vector.
        """
        # Extract rotation part
        rotation_matrix = transformation[:3, :3]
        translation_vector = transformation[:3, 3]

        # Convert rotation matrix to quaternion
        quaternion = self.rotationMatrixToQuaternion(rotation_matrix)

        return quaternion, translation_vector
    
    def rotationMatrixToQuaternion(self, rotation_matrix: np.ndarray) -> np.ndarray:
        """
        Converts a rotation matrix into a quaternion representation.

        :param rotation_matrix: A 3x3 numpy array representing the rotation matrix.
        :return: A numpy array representing the quaternion in the form [x, y, z, w].
        """
        # Check if the input is a valid rotation matrix
        if not np.allclose(np.dot(rotation_matrix, rotation_matrix.T), np.eye(3), atol=1e-6) or not np.isclose(np.linalg.det(rotation_matrix), 1.0, atol=1e-6):
            raise ValueError("Input matrix is not a valid rotation matrix.")
        
        # Calculate the trace of the rotation matrix
        trace = np.trace(rotation_matrix)

        # If the trace is greater than zero, use the first method to calculate the quaternion
        if trace > 0:
            w = np.sqrt(1.0 + trace) / 2.0  # Calculate the scalar part (w) of the quaternion
            x = (rotation_matrix[2, 1] - rotation_matrix[1, 2]) / (4.0 * w)  # Calculate x component
            y = (rotation_matrix[0, 2] - rotation_matrix[2, 0]) / (4.0 * w)  # Calculate y component
            z = (rotation_matrix[1, 0] - rotation_matrix[0, 1]) / (4.0 * w)  # Calculate z component
        else:
            # If the trace is not positive, use the largest diagonal element to determine the dominant axis
            if rotation_matrix[0, 0] > rotation_matrix[1, 1] and rotation_matrix[0, 0] > rotation_matrix[2, 2]:
                # Calculate x component based on the first row/column
                x = np.sqrt(1.0 + rotation_matrix[0, 0] - rotation_matrix[1, 1] - rotation_matrix[2, 2]) / 2.0
                w = (rotation_matrix[2, 1] - rotation_matrix[1, 2]) / (4.0 * x)  # Calculate w component
                y = (rotation_matrix[0, 1] + rotation_matrix[1, 0]) / (4.0 * x)  # Calculate y component
                z = (rotation_matrix[0, 2] + rotation_matrix[2, 0]) / (4.0 * x)  # Calculate z component
            elif rotation_matrix[1, 1] > rotation_matrix[2, 2]:
                # Calculate y component based on the second row/column
                y = np.sqrt(1.0 + rotation_matrix[1, 1] - rotation_matrix[0, 0] - rotation_matrix[2, 2]) / 2.0
                w = (rotation_matrix[0, 2] - rotation_matrix[2, 0]) / (4.0 * y)  # Calculate w component
                x = (rotation_matrix[0, 1] + rotation_matrix[1, 0]) / (4.0 * y)  # Calculate x component
                z = (rotation_matrix[1, 2] + rotation_matrix[2, 1]) / (4.0 * y)  # Calculate z component
            else:
                # Calculate z component based on the third row/column
                z = np.sqrt(1.0 + rotation_matrix[2, 2] - rotation_matrix[0, 0] - rotation_matrix[1, 1]) / 2.0
                w = (rotation_matrix[1, 0] - rotation_matrix[0, 1]) / (4.0 * z)  # Calculate w component
                x = (rotation_matrix[0, 2] + rotation_matrix[2, 0]) / (4.0 * z)  # Calculate x component
                y = (rotation_matrix[1, 2] + rotation_matrix[2, 1]) / (4.0 * z)  # Calculate y component

        # Return the quaternion as a numpy array in the form [x, y, z, w]
        return np.array([x, y, z, w])

    
    def createTransformationMatrix(self, quaternion: np.ndarray, translation: np.ndarray) -> np.ndarray:
        """
        Creates a 4x4 transformation matrix from a quaternion and a translation vector.

        :param quaternion: A numpy array representing the quaternion [x, y, z, w].
        :param translation: A numpy array representing the translation vector [tx, ty, tz].
        :return: A 4x4 transformation matrix.
        """
        x, y, z, w = quaternion

        # Create rotation matrix from quaternion
        rotation_matrix = np.array([
            [1 - 2 * (y**2 + z**2), 2 * (x * y - z * w), 2 * (x * z + y * w)],
            [2 * (x * y + z * w), 1 - 2 * (x**2 + z**2), 2 * (y * z - x * w)],
            [2 * (x * z - y * w), 2 * (y * z + x * w), 1 - 2 * (x**2 + y**2)]
        ])

        # Create the transformation matrix
        transformation_matrix = np.eye(4)
        transformation_matrix[:3, :3] = rotation_matrix
        transformation_matrix[:3, 3] = translation

        return transformation_matrix

File: src/robot_control.cpp
#include "robot_control.hpp"

// Constructor: Initializes publishers and subscribers, and sets control parameters
UR5Control::UR5Control(const ros::NodeHandle& nh, double freq,
                       const motion_profile_settings_t &joints_profile, 
                       const cart_motion_profile_settings_t &cart_profile) :
    nh_(nh), freq_(freq),
    current_joints_vel_(Vector6d::Zero()),
    joints_profile_(joints_profile), 
    cart_profile_(cart_profile) 
{
    joint_state_subscriber_ = nh_.subscribe("/joint_states", 1, &UR5Control::jointStateCallback, this);
    joint_position_pub_ = nh_.advertise<std_msgs::Float64MultiArray>("/joint_group_position_controller/command", 1);
}

// Callback: Handles incoming joint state messages and updates the joint angles and velocities
void UR5Control::jointStateCallback(const sensor_msgs::JointState::ConstPtr& msg) {
    Vector6d joints_pos;
    if (msg->position.size() == joints_pos.size()) { 
        for (size_t i = 0; i < joints_pos.size(); ++i) {
            joints_pos[i] = msg->position[i];
            current_joints_vel_[i] = msg->velocity[i];
        }
        //joint_angles_ = wrapToPi(joints_pos);
    } else {
        ROS_WARN("Received joint state with insufficient position data.");
    }
}

// Sets the desired joint positions by publishing to the joint position topic
void UR5Control::setJointPos(const Vector6d& joints_pos) {
    std_msgs::Float64MultiArray joint_positions; // Create a message to hold joint positions
    setJointAngles(joints_pos); // Update the internal joint angles with the new positions
    joint_positions.data.resize(6); // Resize the message array to hold 6 joint positions
    joint_positions.data.assign(joints_pos.data(), joints_pos.data() + joints_pos.size()); // Copy joint positions into the message
    joint_position_pub_.publish(joint_positions); // Publish the joint positions to the ROS topic
}

// Moves the robot to a target position using a point-to-point motion profile
bool UR5Control::movePtP(const Matrix4d& target_transform, const motion_profile_settings_t& profile_settings) {
    Matrix86d joint_solutions; // Storage for the inverse kinematics solutions
    int num_solutions = inverseKinematics(target_transform, joint_solutions); // Calculate inverse kinematics for target position
    if (num_solutions <= 0) { // Check if any solutions were found
        ROS_ERROR("Target position is not reachable"); // Log an error if not reachable
        return false; // Return false to indicate failure
    }

    // Find the nearest joint angles solution from the computed IK solutions
    Vector6d target_joints_angles = findNearestIKSolution(joint_angles_, joint_solutions);
    Ruckig<6> otg(1./freq_); // Initialize the trajectory generator with frequency
    InputParameter<6> input; // Create input parameters for motion
    OutputParameter<6> output; // Create output parameters for motion

    // Initialize input parameters for motion profile
    input.current_velocity.fill(0);
    input.current_acceleration.fill(0);
    input.target_velocity.fill(0);
    input.target_acceleration.fill(0);
    std::copy(joint_angles_.data(), joint_angles_.data() + 6, input.current_position.begin()); // Set current joint positions
    std::copy(target_joints_angles.data(), target_joints_angles.data() + 6, input.target_position.begin()); // Set target joint positions

    // Set motion profile limits
    input.max_velocity.fill(profile_settings.max_velocity);
    input.max_acceleration.fill(profile_settings.max_acceleration);
    input.max_jerk.fill(profile_settings.jerk);

    ros::Rate rate(freq_); // Set loop rate based on frequency
    while (otg.update(input, output) == Result::Working && ros::ok()) { // Update motion until completed or interrupted
        setJointPos(Map<Vector6d>(output.new_position.data(), 6)); // Set the new joint positions
        output.pass_to_input(input); // Pass output parameters to input for the next iteration
        rate.sleep(); // Sleep for the defined rate
        ros::spinOnce(); // Allow ROS to process incoming messages
    }
    return true; // Return true to indicate successful movement
}

// Moves the robot in a linear trajectory from the current to the target pose
bool UR5Control::moveLine(const Matrix4d& target_transform, const cart_motion_profile_settings_t& profile_settings) {
    Matrix86d target_solutions, current_solutions; // Storage for target and current IK solutions
    // Calculate inverse kinematics for target position
    if (inverseKinematics(target_transform, target_solutions) == 0) {
        ROS_ERROR("Target position is not reachable"); // Log an error if not reachable
        return false; // Return false to indicate failure
    }

    Vector6d initial_joint_angles = getJointAngles(); // Get the current joint angles
    Matrix4d current_transform = forwardKinematics(initial_joint_angles); // Calculate the current transformation from joint angles
    // Calculate inverse kinematics for the current position
    if (inverseKinematics(current_transform, current_solutions) == 0) {
        ROS_ERROR("Current position is invalid"); // Log an error if the current position is invalid
        return false; // Return false to indicate failure
    }

    int n_sol = 0; // Variable to hold the index of the valid solution
    for (int i = 0; i < current_solutions.rows(); ++i) { // Find the closest solution to the initial joint angles
        if (current_solutions.row(i).isApprox(initial_joint_angles.transpose(), 1e-5)) {
            n_sol = i; // Store the index of the matching solution
            break; // Exit loop upon finding the match
        }
    }

    // Check if the solution is valid
    if (target_solutions.row(n_sol).isZero() || target_solutions.row(n_sol).hasNaN()) {
        ROS_ERROR("Cannot generate linear motion"); // Log an error if linear motion cannot be generated
        return false; // Return false to indicate failure
    }

    InputParameter<7> input; // Create input parameters for motion with 7 dimensions
    OutputParameter<7> output; // Create output parameters for motion
    Ruckig<7> otg(1./freq_); // Initialize trajectory generator for 7 dimensions
    Quaterniond target_quat, current_quat; // Quaternions for target and current orientation
    Vector3d target_pos, current_pos; // Vectors for target and current position

    // Extract rotation and translation from the target and current transformation matrices
    extractRotationAndTranslation(target_transform, target_quat, target_pos);
    extractRotationAndTranslation(current_transform, current_quat, current_pos);
    // Initialize input parameters for motion profile
    input.current_velocity.fill(0);
    input.current_acceleration.fill(0);
    input.target_velocity.fill(0);
    input.target_acceleration.fill(0);
    // Set current and target positions and orientations in the input
    std::copy(current_quat.coeffs().data(), current_quat.coeffs().data() + 4, input.current_position.begin());
    std::copy(target_quat.coeffs().data(), target_quat.coeffs().data() + 4, input.target_position.begin());
    std::copy(current_pos.data(), current_pos.data() + 3, input.current_position.begin() + 4);
    std::copy(target_pos.data(), target_pos.data() + 3, input.target_position.begin() + 4);

    // Set maximum velocity, acceleration, and jerk for angular motion
    for (size_t i = 0; i < 4; ++i) {
        input.max_velocity[i] = profile_settings.angular.max_velocity;
        input.max_acceleration[i] = profile_settings.angular.max_acceleration;
        input.max_jerk[i] = profile_settings.angular.jerk;
    }
    // Set maximum velocity, acceleration, and jerk for linear motion
    for (size_t i = 4; i < 7; ++i) {
        input.max_velocity[i] = profile_settings.linear.max_velocity;
        input.max_acceleration[i] = profile_settings.linear.max_acceleration;
        input.max_jerk[i] = profile_settings.linear.jerk;
    }

    ros::Rate rate(freq_); // Set loop rate based on frequency
    while (otg.update(input, output) == Result::Working && ros::ok()) { // Update motion until completed or interrupted
        Quaterniond new_quat(Map<Vector4d>(output.new_position.data())); // Extract new quaternion from output
        new_quat.normalize(); // Normalize the quaternion
        Vector3d new_pos = Map<Vector3d>(output.new_position.data() + 4); // Extract new position from output

        // Create a new transformation matrix from the new quaternion and position
        Matrix4d new_transform = createTransformationMatrix(new_quat, new_pos);
        Matrix86d joint_solutions; // Storage for joint solutions
        int num_solutions = inverseKinematics(new_transform, joint_solutions); // Calculate IK for the new transform
        if (num_solutions <= 0) {
            ROS_ERROR("Target position is not reachable"); // Log an error if not reachable
            return false; // Return false to indicate failure
        }
        // Find the nearest joint angles solution from the computed IK solutions
        Vector6d target_joints_angles = findNearestIKSolution(getJointAngles(), joint_solutions);
        setJointPos(target_joints_angles); // Set the joint positions
        output.pass_to_input(input); // Pass output parameters to input for the next iteration
        rate.sleep(); // Sleep for the defined rate
        ros::spinOnce(); // Allow ROS to process incoming messages
    }
    return true; // Return true to indicate successful linear movement
}

bool UR5Control::movePtP(const Matrix4d& target_transform)
{
    return movePtP(target_transform, joints_profile_);
}

bool UR5Control::moveLine(const Matrix4d& target_transform)
{
    return moveLine(target_transform, cart_profile_);
}